pkg/container/qemu_runner.go

```diff
@@ -0,0 +1,879 @@
+					clog.FromContext(ctx).Errorf("qemu: can't unpack kernel - %v", err)
+					return "", err
+				}
+			case tar.TypeReg:
+				// Create file
+				outFile, err := os.Create(target)
+				if err != nil {
+					clog.FromContext(ctx).Errorf("qemu: can't unpack kernel - %v", err)
+					return "", err
+				}
+				defer outFile.Close()
+
+				if _, err := io.Copy(outFile, tarReader); err != nil {
+					clog.FromContext(ctx).Errorf("qemu: can't unpack kernel - %v", err)
+					return "", err
+				}
+			default:
+				clog.FromContext(ctx).Warnf("Unsupported type: %v in %s\n", header.Typeflag, header.Name)
+			}
+		}
+	}
+
+	clog.FromContext(ctx).Info("qemu: detecting default kernel in /boot/vmlinuz")
+	if _, err := os.Stat("/boot/vmlinuz"); err == nil {
+		clog.FromContext(ctx).Info("qemu: /boot/vmlinuz detected")
+		return "/boot/vmlinuz", nil
+	}
+
+	return "", fmt.Errorf("no kernel specified, could not find default /boot/vmlinuz")
+}
+
+func generateInitrd(ctx context.Context, rootfs string) (string, error) {
+	clog.FromContext(ctx).Info("qemu: building initramfs...")
+
+	findFileOutput, err := os.CreateTemp("", "")
+	if err != nil {
+		clog.FromContext(ctx).Errorf("qemu: rootfs file fidning output file failed: %v", err)
+		return "", err
+	}
+	defer findFileOutput.Close()
+	defer os.Remove(findFileOutput.Name())
+
+	findFiles := exec.Command("find", ".")
+	findFiles.Dir = rootfs
+	findFiles.Stdout = findFileOutput
+
+	clog.FromContext(ctx).Debug("qemu: finding rootfs files...")
+	err = findFiles.Run()
+	if err != nil {
+		clog.FromContext(ctx).Errorf("qemu: rootfs file fidning failed: %v", err)
+		return "", err
+	}
+
+	initramfs, err := os.Create(filepath.Join(rootfs, "initramfs.cpio"))
+	if err != nil {
+		clog.FromContext(ctx).Errorf("qemu: initramfs file creation failed: %v", err)
+		return "", err
+	}
+	defer initramfs.Close()
+
+	content, _ := os.ReadFile(findFileOutput.Name())
+	buffer := bytes.Buffer{}
+	buffer.Write(content)
+
+	cpioCmd := exec.Command("cpio", "--format=newc", "-o", "-R", "0:0")
+
+	cpioCmd.Stdout = initramfs
+	cpioCmd.Stdin = &buffer
+	cpioCmd.Dir = rootfs
+
+	clog.FromContext(ctx).Info("qemu: compressing rootfs...")
+	clog.FromContext(ctx).Debugf("qemu: launching command - %s",
+		strings.Join(cpioCmd.Args, " ")+" | "+strings.Join(cpioCmd.Args, " "))
+	if err := cpioCmd.Run(); err != nil {
+		clog.FromContext(ctx).Errorf("qemu: initramfs cpio command failed: %v", err)
+		return "", err
+	}
+
+	clog.FromContext(ctx).Info("qemu: initramfs ready")
+	return initramfs.Name(), nil
+}
+
+func sendSSHCommand(ctx context.Context, user, host, port string,
+	cfg *Config, extraVars map[string]string,
+	stdin io.Reader, stderr, stdout io.Writer,
+	command []string,
+) error {
+	server := host + ":" + port
+
+	signer, err := ssh.ParsePrivateKey(cfg.SSHKey)
+	if err != nil {
+		clog.FromContext(ctx).Errorf("Unable to parse private key: %v", err)
+		return err
+	}
+
+	// Create SSH client configuration
+	config := &ssh.ClientConfig{
```

Comment:
We could specify the cipher and potentially increase the ssh performance. [this](https://www.linkedin.com/pulse/bench-marking-ssh-ciphers-transfer-speed-phil-cryer-eakxc/) says that aes128-ctr is the fastest of those enabled, and reports it is better than 2x the performance of the default (chacha20).

Reply:
Thanks @smoser addressed this :+1: